wgsl/index.bs

```diff
@@ -353,59 +353,102 @@ Note: Only a function declaration can contain other declarations.
 
 # Types # {#types}
 
-Note: For the syntax of declaring types in [SHORTNAME] please see the [[#grammar]].
-TODO(dneto): This note is probably editorially obsolete.
+Programs calculate values.
 
-Programs calculate values. Each value in [SHORTNAME] belongs to exactly one *type*.
-A type is a set of (mathematical) values.
+In [SHORTNAME], a <dfn noexport>type</dfn> is set of values, and each value belongs to exactly one type.
+A value's type determines the syntax and semantics of operations that can be performed on that value.
 
-We distinguish between the concept of a type and the syntax in [SHORTNAME] to denote that type.
-In many cases the spelling of a type in this document is the same as its [SHORTNAME] syntax.
-The spelling is different for structure types, or types containing structures.
+A type is either predeclared, or created in WGSL source via a [=declaration=].
 
-## Type Checking ## {#type-checking}
+We distinguish between the *concept* of a type and the *syntax* in [SHORTNAME] to denote that type.
+In many cases the spelling of a type in this specification is the same as its [SHORTNAME] syntax.
+For example:
+* the set of 32-bit unsigned integer values is spelled `u32` in this specification,
+    and also in a [SHORTNAME] program.
+* the spelling is different for structure types, or types containing structures.
 
-Type checking is the process of mapping terms in the [SHORTNAME] source language to [[#types]].
+Some [SHORTNAME] types are only used for analyzing a source program and
+for determining the program's runtime behaviour.
+This specification will describe such types, but they do not appear in [SHORTNAME] source text.
 
-Generally, we start by determining types for the smallest [SHORTNAME] source phrases, and then build up
-via combining rules.
+Note: [SHORTNAME] reference types are not written in [SHORTNAME] programs. See TODO forward reference to ptr/ref.
 
-If we can derive a type for the whole [SHORTNAME] source program via the type rules, then we say
-the program is *well-typed*.  Otherwise there is a type error and is not a valid
-[SHORTNAME] program.
+## Type Checking ## {#type-checking-section}
 
+A [SHORTNAME] value is computed by evaluating an expression.
+An <dfn noexport>expression</dfn> is a segment of source text
+parsed as one of the [SHORTNAME] grammar rules whose name ends with "`_expression`".
+An expression *E* can contain <dfn noexport>sub-expressions</dfn> which are expressions properly contained
+in the outer expression *E*.
 
-Issue: (dneto) complete
+The particular value produced by an expression evaluation depends on:
+* <dfn noexport>static context</dfn>:
+    the source text surrounding the expression, and
+* <dfn noexport>dynamic context</dfn>:
+    the state of the invocation evaluating the expression,
```

Comment:
should we link to the definition of "invocation" here?

Reply:
Looks like we don't have a definition of "invocation" yet.
I filed #1555 to make one.